File: collector/core/message.go
package core

// Message is a single collected record: a structured payload plus metadata
// describing where it came from.
type Message struct {
	Data map[string]any    `json:"data"`
	Meta map[string]string `json:"meta"`
}

// NewMessage returns a Message with an initialised metadata map.
func NewMessage(data map[string]any) Message {
	return Message{Data: data, Meta: map[string]string{}}
}

File: collector-rs/src/message.rs
use std::collections::BTreeMap;

use serde::Serialize;

/// A single collected record: a structured payload plus metadata describing
/// where it came from.
#[derive(Debug, Clone, Serialize)]
pub struct Message {
    pub data: serde_json::Value,
    pub meta: BTreeMap<String, String>,
}

impl Message {
    pub fn new(data: serde_json::Value) -> Self {
        Self {
            data,
            meta: BTreeMap::new(),
        }
    }

    pub fn with_meta(mut self, key: &str, value: impl Into<String>) -> Self {
        self.meta.insert(key.to_string(), value.into());
        self
    }
}
